File: tools/gen-events.mjs
#!/usr/bin/env node
/*
 * 이벤트 일러스트 자동 생성기 (OpenAI Images / gpt-image-1)
 * 삼국지(코에이 RoTK) 이벤트 CG풍 — 시대별 역사 이벤트 장면.
 *
 * 사용법:
 *   export OPENAI_API_KEY=sk-...
 *   npm i sharp                                  # 3:2 webp 변환(권장)
 *   node tools/gen-events.mjs                     # 없는 것만 생성 + EVENT_IMG 자동 등록
 *   node tools/gen-events.mjs --only u_ansi,g_pyeongyang
 *   node tools/gen-events.mjs --force             # 이미 있어도 다시 생성
 *   node tools/gen-events.mjs --dry-run           # 호출 없이 프롬프트만 출력
 *   node tools/gen-events.mjs --register-only     # 폴더 스캔해 EVENT_IMG만 갱신
 * 옵션 환경변수: OPENAI_IMAGE_MODEL(기본 gpt-image-1), IMAGE_QUALITY(low|medium|high), GEN_SIZE(기본 1536x1024)
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dir = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dir, '..');
const INDEX = path.join(ROOT, 'index.html');
const OUTDIR = path.join(ROOT, 'assets', 'events');

const args = process.argv.slice(2);
const has = f => args.includes(f);
const val = f => { const i = args.indexOf(f); return i >= 0 ? args[i + 1] : null; };
const DRY = has('--dry-run'), FORCE = has('--force'), REGISTER_ONLY = has('--register-only');
const ONLY = (val('--only') || '').split(',').map(s => s.trim()).filter(Boolean);

const MODEL = process.env.OPENAI_IMAGE_MODEL || 'gpt-image-1';
const QUALITY = process.env.IMAGE_QUALITY || 'high';
const SIZE = process.env.GEN_SIZE || '1536x1024';   // 가로형 3:2
const OUT_W = 960, OUT_H = 640;

const COMMON = `A dramatic cinematic historical scene illustration of the Korean Three Kingdoms era, in the style of Koei Romance of the Three Kingdoms event CG art, semi-realistic digital painting, epic atmosphere, dynamic composition, dramatic cinematic lighting, period-accurate lamellar armor and banners, painterly detail, 3:2 landscape, no text, no watermark, no border, do NOT resemble any modern actor or copyrighted artwork`;

/* 이벤트별 장면 묘사(고증·설화 기반) */
const EVT_EN = {
  // 371 근초고왕
  g_intro:    'King Geunchogo of Baekje at the height of his power reviewing his great army, golden Baekje banners, a prosperous fortress capital behind, confident and majestic, warm dawn light',
  g_pyeongyang:'Baekje army storming the stone walls of Pyeongyang fortress at dusk, scaling ladders and arrows, the Goguryeo king Gogugwon struck by an arrow and falling among his guards, fire and chaos',
  g_chiljido: 'a solemn ceremony where Baekje envoys present the famous seven-branched sword (Chiljido) to Wa Yamato envoys, ornate hall, gold and crimson, diplomacy and alliance',
  g_sosurim:  'King Sosurim of Goguryeo proclaiming new state laws and founding a royal academy, Buddhist monks and Confucian scholars present, candlelit hall, scene of reform and revival',
  g_mahan:    'Baekje army subjugating the Mahan chiefdoms across the southern plains, kneeling local lords offering submission, rows of Baekje banners, autumn fields',
  // 400 광개토대왕
  w_intro:    'the young Gwanggaeto the Great enthroned as king of Goguryeo, ornate gold crown over a feathered cap, mighty broad-shouldered presence, golden throne hall, rays of light',
  w_baekje:   'Goguryeo heavy cavalry overwhelming the Baekje army at a river, the Baekje king Asin kneeling in submission before mounted Gwanggaeto, banners and dust',
  w_namjeong: 'Goguryeo heavy cavalry (gaemamusa, armored horse riders) crushing Wa raiders and Gaya forces along the Nakdong River to rescue Silla, spears and broken enemies, year 400',
  w_yodong:   'Goguryeo army storming the Liaodong plain and driving out the Later Yan Xianbei cavalry, a great fortress aflame on the horizon, victorious advance',
  w_buyeo:    'Goguryeo army on the northern steppe receiving the submission of Eastern Buyeo, vast columns of soldiers and banners under a wide sky',
  w_hanseong: 'King Jangsu of Goguryeo leading a vast army assaulting Wiryeseong, the Baekje royal capital falling in flames, the captured Baekje king Gaero, somber and epic',
  // 554 진흥왕
  j_intro:    'Silla King Jinheung and Baekje allies sealing the Na-Je alliance pact in a hall overlooking the Han River, banners of both kingdoms, tense diplomacy',
  j_hangang:  'Silla army seizing the lower Han River region and betraying the Baekje alliance, Silla banners planted by the great river at dawn',
  j_gwansan:  'the Battle of Gwansan fortress, the Baekje king Seong ambushed and slain by Silla soldiers in a narrow valley, somber dramatic dusk, broken Baekje army',
  j_daegaya:  'Silla generals Isabu and Sadaham conquering Daegaya, the last Gaya kingdom, its palace falling, iron-armored Gaya warriors defeated, melancholy',
  j_sunsubi:  'King Jinheung of Silla erecting a tall stone monument (sunsubi) on a windswept mountain peak, surveying his vastly expanded realm, banners and clouds',
  // 645 여당전쟁
  u_decree:   'Tang Emperor Taizong personally leading a vast Tang army crossing the Liao River toward Goguryeo, endless ranks of banners and siege engines, overwhelming might',
  u_war1:     'the Tang army crossing the Liao River and assaulting Goguryeo border fortresses, massive siege with catapults and ladders, smoke over the walls',
  u_ansi:     'the heroic siege of Ansi fortress, commander Yang Manchun and Goguryeo soldiers defending the stone walls against a giant Tang earthwork siege ramp, arrows and defiance',
  u_retreat:  'the Tang army retreating through winter snow after failing to take Ansi fortress, weary defeated columns, Goguryeo defenders victorious on the walls at dawn',
  u_fall:     'Goguryeo Liaodong defenses collapsing as the Tang army advances, fortresses burning on the plain, retreating Goguryeo soldiers, somber and grim',
};
function buildPrompt(id){ return `${COMMON}. Scene: ${EVT_EN[id]}.`; }

async function saveImage(buf, name) {
  fs.mkdirSync(OUTDIR, { recursive: true });
  let sharp = null; try { sharp = (await import('sharp')).default; } catch (_) {}
  if (sharp) { const out = path.join(OUTDIR, `${name}.webp`);
    await sharp(buf).resize(OUT_W, OUT_H, { fit: 'cover' }).webp({ quality: 82 }).toFile(out);
    return path.relative(ROOT, out); }
  const out = path.join(OUTDIR, `${name}.png`); fs.writeFileSync(out, buf);
  console.warn('  ⚠ sharp 미설치 → png 저장(권장: npm i sharp)'); return path.relative(ROOT, out);
}
async function generate(prompt) {
  const key = process.env.OPENAI_API_KEY; if (!key) throw new Error('OPENAI_API_KEY 환경변수가 없습니다.');
  for (let a = 1; a <= 5; a++) {
    const res = await fetch('https://api.openai.com/v1/images/generations', {
      method: 'POST', headers: { 'Authorization': `Bearer ${key}`, 'Content-Type': 'application/json' },
      body: JSON.stringify({ model: MODEL, prompt, size: SIZE, quality: QUALITY, n: 1 }) });
    if (res.ok) { const j = await res.json(); const b64 = j.data?.[0]?.b64_json;
      if (!b64) throw new Error('이미지 데이터 없음: ' + JSON.stringify(j).slice(0, 200)); return Buffer.from(b64, 'base64'); }
    if (res.status === 429 || res.status >= 500) { const w = 2 ** a; console.warn(`  재시도 ${a} (HTTP ${res.status}) ${w}s`); await new Promise(r => setTimeout(r, w * 1000)); continue; }
    throw new Error(`HTTP ${res.status}: ${(await res.text()).slice(0, 300)}`);
  }
  throw new Error('재시도 초과');
}
function registerAll() {
  const files = fs.existsSync(OUTDIR) ? fs.readdirSync(OUTDIR).filter(f => /\.(webp|png|jpg|jpeg)$/i.test(f)) : [];
  const entries = files.map(f => ({ id: f.replace(/\.(webp|png|jpg|jpeg)$/i, ''), p: `assets/events/${f}` })).sort((a, b) => a.id.localeCompare(b.id));
  const lines = entries.map(e => `  ${e.id}:'${e.p}',`).join('\n');
  const block = `const EVENT_IMG={\n${lines}\n};`;
  let html = fs.readFileSync(INDEX, 'utf8');
  const re = /const EVENT_IMG=\{[\s\S]*?\n\};/;
  if (!re.test(html)) throw new Error('index.html에서 EVENT_IMG 블록을 찾지 못했습니다.');
  fs.writeFileSync(INDEX, html.replace(re, block));
  console.log(`EVENT_IMG 등록 완료: ${entries.length}개`);
}
async function main() {
  if (REGISTER_ONLY) { registerAll(); return; }
  let ids = Object.keys(EVT_EN); if (ONLY.length) ids = ids.filter(i => ONLY.includes(i));
  console.log(`대상 이벤트: ${ids.length}개 / 모델 ${MODEL} / ${SIZE} / ${QUALITY}${DRY ? ' (DRY-RUN)' : ''}`);
  let made = 0, skip = 0, fail = 0;
  for (const id of ids) {
    const exists = ['webp', 'png', 'jpg', 'jpeg'].some(e => fs.existsSync(path.join(OUTDIR, `${id}.${e}`)));
    if (exists && !FORCE) { skip++; continue; }
    const prompt = buildPrompt(id);
    if (DRY) { console.log(`- ${id}\n    ${prompt}`); continue; }
    try { process.stdout.write(`生成 ${id} ... `); const buf = await generate(prompt); console.log('✓ ' + await saveImage(buf, id)); made++; }
    catch (e) { console.log('✗ ' + e.message); fail++; }
  }
  if (!DRY) { registerAll(); console.log(`\n완료 — 생성 ${made}, 건너뜀 ${skip}, 실패 ${fail}`); }
}
export { EVT_EN, buildPrompt };
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) main().catch(e => { console.error(e); process.exit(1); });

File: scripts/gen_events.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
삼한전 이벤트 이미지 생성기 (gpt-image-1, 레퍼런스 참조)

각 이벤트마다 인물 초상화 + 세력 갑옷 레퍼런스를 함께 첨부(images.edit)해
얼굴·복식을 게임과 통일한 이벤트 일러스트를 만든 뒤, 960x640 webp로 변환해
assets/events/<id>.webp 로 저장한다. EVENT_IMG 매핑(index.html)은 이미 21개가
연결돼 있으므로 파일만 교체되면 즉시 게임에 반영된다.

사용법 (레포 루트에서):
    pip install openai pillow
    set OPENAI_API_KEY=sk-...        (Windows)   /  export OPENAI_API_KEY=sk-...  (mac/linux)
    python scripts/gen_events.py                  # 이미 있는 파일은 건너뜀
    python scripts/gen_events.py --force          # 전부 다시 생성
    python scripts/gen_events.py --only g_intro u_ansi   # 특정 이벤트만
    python scripts/gen_events.py --keep-raw       # 원본 PNG도 assets/_raw_events/ 에 보관

옵션:
    --model    기본 gpt-image-1
    --size     기본 1536x1024 (생성 크기)
    --quality  기본 medium  (low|medium|high|auto)
    --width    기본 960     (저장 webp 가로, 3:2로 리사이즈)
    --wq       기본 82      (webp 품질)
"""
import argparse, base64, io, os, sys, time
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
EVENTS_DIR = ROOT / "assets" / "events"
RAW_DIR = ROOT / "assets" / "_raw_events"
P = lambda *a: str(ROOT.joinpath(*a))   # 레포 기준 경로 헬퍼

COMMON = (
    "Use the attached images as the exact reference for the characters' faces and for the "
    "armor/costume and banner style — keep the same facial features, hairstyle, crown/helmet, "
    "and clothing as in the references. Then create a single illustration: painterly cinematic "
    "historical scene, Korean Three Kingdoms era, dramatic chiaroscuro lighting, muted earthy "
    "palette with gold and crimson accents, semi-realistic game key-art, 3:2 landscape, "
    "atmospheric haze. No text, no letters, no watermark, no modern objects. SCENE: "
)

# (id, refs[], scene)
EVENTS = [
    ("g_intro", ["assets/portraits/baekje/근초고왕.webp", "assets/refs/baekje/백제갑옷.png"],
     "A Baekje king (the man in the reference) in gilt-bronze crown and ornate scale armor stands on a "
     "hilltop palace terrace at golden dawn, overlooking a fertile river plain with thatched villages and "
     "ships on the sea horizon, purple royal banners in the wind, hand resting on his sword toward the north."),
    ("g_pyeongyang", ["assets/portraits/baekje/근구수왕.webp", "assets/portraits/goguryeo/고국원왕.webp"],
     "Night siege of a Goguryeo mountain fortress. Baekje troops in scale armor storm the gate under purple "
     "banners; in the foreground the Goguryeo king (reference face) in brown lamellar reels backward struck "
     "by an arrow in the chest, his green banner falling. Fire, arrows, embers. Tragic decisive moment."),
    ("g_chiljido", ["assets/refs/baekje/칠지도.png", "assets/portraits/wa/오토모.webp"],
     "A diplomatic hall: a Baekje envoy in purple robes presents the iron Seven-Branched Sword (a straight "
     "blade with three branching prongs on each side, per reference) on a lacquer tray to a seated Wa "
     "chieftain (reference face) in simple short armor. Teal banners, incense smoke, warm lamplight."),
    ("g_sosurim", ["assets/portraits/goguryeo/소수림왕.webp", "assets/refs/goguryeo/고구려 갑옷03.png"],
     "The Goguryeo king (reference face) in dark lamellar and a feather-cap receives a saffron-robed Buddhist "
     "monk in a new temple hall, scholars unrolling bamboo law-code scrolls, a bronze Buddha glowing in "
     "candlelight. Green banners. Rebuilding, enlightenment, quiet resolve."),
    ("g_mahan", ["assets/portraits/baekje/근초고왕.webp"],
     "Baekje cavalry and infantry in scale armor march south through humid green wetlands and reed fields, "
     "receiving the submission of kneeling Mahan village chiefs, purple banners, rice paddies, low golden light."),

    ("w_intro", ["assets/portraits/goguryeo/광개토대왕.webp", "assets/refs/goguryeo/고구려 갑옷03.png"],
     "The young Goguryeo king (reference face, age 18) crowned, in gleaming brown lamellar and a plumed "
     "feather-cap, raising a sword skyward; massed cataphract cavalry and green banners behind him under a "
     "vast dawn sky. Youthful conquering ambition."),
    ("w_baekje", ["assets/portraits/goguryeo/광개토대왕.webp", "assets/portraits/baekje/아신왕.webp"],
     "By a wide river after battle: the defeated Baekje king (reference face) in dented scale armor kneels in "
     "the mud in surrender before the towering Goguryeo king (reference) on an armored warhorse; captured "
     "purple banners trail in the dirt, victorious green banners rise, ash in the air."),
    ("w_namjeong", ["assets/portraits/goguryeo/광개토대왕.webp", "assets/refs/gaya/가야갑옷02.png"],
     "Goguryeo heavy cataphracts (gaema musa, full iron horse-and-rider armor) charge through a river ford, "
     "routing lightly-armored Wa warriors with curved swords and crimson iron-plate-armored Gaya troops; "
     "spray of water, broken teal and crimson banners, dawn light."),
    ("w_yodong", ["assets/portraits/goguryeo/광개토대왕.webp"],
     "Goguryeo army assaults a great Liaodong plains fortress at dusk: brown-lamellar troops scale walls with "
     "ladders while cataphracts ride down fleeing Later-Yan Chinese soldiers in red; vast flat horizon, smoke "
     "pillars, green banners planted on the ramparts."),
    ("w_buyeo", ["assets/portraits/goguryeo/광개토대왕.webp"],
     "A Goguryeo column marches across a frozen northern snowfield toward a palisaded Buyeo settlement; breath "
     "fogs in the cold air, green banners stiff with frost, the king on horseback surveying distant captured "
     "forts. Pale blue winter light."),
    ("w_hanseong", ["assets/portraits/goguryeo/고거련.webp"],
     "A Baekje royal capital burns at night; the Goguryeo king Jangsu (reference face) in imperial lamellar "
     "watches from a ridge as his army storms the walls; in the foreground a captured Baekje king is led away "
     "in chains under green banners. Heavy smoke, collapsing gate, tragic grandeur."),

    ("j_intro", ["assets/portraits/silla/진흥왕.webp", "assets/portraits/baekje/성왕.webp"],
     "Two allied kings — the Silla king (reference) in a radiant gold crown and the Baekje king (reference) in "
     "a gilt-bronze crown — clasp arms before a battle map table, gold and purple banners side by side, a "
     "flicker of hidden ambition in the Silla king's eyes. Warm war-tent lamplight."),
    ("j_hangang", ["assets/portraits/silla/진흥왕.webp", "assets/portraits/baekje/성왕.webp"],
     "Silla cavalry in gilt armor under gold banners launch a surprise dawn attack to seize fortresses along "
     "the lower Han River, planting a new gold banner on a captured wall; in the distance the Baekje king "
     "(reference) watches trembling with rage, his purple banner cut down."),
    ("j_gwansan", ["assets/portraits/baekje/성왕.webp", "assets/portraits/silla/김무력.webp"],
     "Ambush at Gwansan fortress: the Baekje king (reference face) is dragged from his horse by Silla soldiers "
     "ambushing from wooded hills, Silla general Kim Muryeok (reference) looking on under gold banners; purple "
     "banners collapse as troops break in panic. Dusk, dust, decisive tragedy."),
    ("j_daegaya", ["assets/portraits/silla/이사부.webp", "assets/portraits/silla/사다함.webp", "assets/portraits/gaya/도설지왕.webp"],
     "The fall of Daegaya: Silla general Isabu (reference) and a young hwarang warrior (reference, floral "
     "feather-cap) breach a crimson-bannered Gaya fortress defended by iron-plate-armored soldiers; a Gaya "
     "court musician clutches a gayageum amid the ruin, gold banners rising over crimson. Elegiac dusk."),
    ("j_sunsubi", ["assets/portraits/silla/진흥왕.webp"],
     "The Silla king (reference) in a gold crown stands atop a windswept mountain peak beside a newly erected "
     "blank stone monument, gazing over a vast realm reaching to distant northern plains, gold banners, eagles, "
     "sweeping clouds. Triumphant dawn."),

    ("u_decree", ["assets/portraits/china/이세민.webp", "assets/portraits/goguryeo/연개소문.webp"],
     "Emperor Tang Taizong (reference face) in Tang bright-armor with round breastplate discs raises his hand "
     "before an endless host crossing the Liao River, countless red dragon banners to the horizon; in the "
     "distant atmosphere a grim Goguryeo dictator Yeon Gaesomun (reference) in black lamellar with twin swords. "
     "Epic scale, dust, war-drums."),
    ("u_war1", ["assets/portraits/china/이세적.webp"],
     "Tang army in red banners and bright-armor storms a Goguryeo wall-fortress with siege ladders and "
     "battering rams; brown-lamellar Goguryeo defenders hurl stones from the ramparts; a long stone "
     "Thousand-li Wall snakes across the hills behind. Smoke, chaos, desperate defense."),
    ("u_ansi", ["assets/portraits/goguryeo/양만춘.webp"],
     "The siege of Ansi fortress: the Goguryeo commander Yang Manchun (reference face) in brown lamellar stands "
     "defiant atop an unbroken stone wall under green banners, a massive Tang-built earthen siege ramp looming "
     "beside the wall, a red sea of Tang troops below, arrows flying. Heroic, indomitable."),
    ("u_retreat", ["assets/portraits/china/이세민.webp", "assets/portraits/goguryeo/양만춘.webp"],
     "Winter retreat: the Tang army trudges away in a snowstorm across the frozen Liao plain, red banners "
     "drooping, supply wagons stuck; the emperor (reference) looks back bitterly at the distant green-bannered "
     "Ansi fortress that never fell. Cold blue light, falling snow."),
    ("u_fall", ["assets/portraits/china/이세적.webp"],
     "Dark fate: Ansi fortress falls, its green banners burning, the Liaodong defensive line collapses; a vast "
     "Tang army in red pours southward through a shattered gate at blood-red sunset, Goguryeo survivors "
     "fleeing. Ominous, candle-in-the-wind peril, heavy crimson sky."),
]


def to_webp(png_bytes: bytes, out_path: Path, width: int, wq: int):
    from PIL import Image
    im = Image.open(io.BytesIO(png_bytes)).convert("RGB")
    h = round(width * 2 / 3)                       # 3:2
    im = im.resize((width, h), Image.LANCZOS)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    im.save(out_path, "WEBP", quality=wq, method=6)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--model", default="gpt-image-1")
    ap.add_argument("--size", default="1536x1024")
    ap.add_argument("--quality", default="medium")
    ap.add_argument("--width", type=int, default=960)
    ap.add_argument("--wq", type=int, default=82)
    ap.add_argument("--force", action="store_true")
    ap.add_argument("--keep-raw", action="store_true")
    ap.add_argument("--only", nargs="*", default=None)
    args = ap.parse_args()

    if not os.environ.get("OPENAI_API_KEY"):
        sys.exit("OPENAI_API_KEY 환경변수를 설정하세요.")
    try:
        from openai import OpenAI
    except ImportError:
        sys.exit("pip install openai pillow 가 필요합니다.")
    client = OpenAI()

    targets = [e for e in EVENTS if (args.only is None or e[0] in args.only)]
    print(f"대상 이벤트: {len(targets)}개 / 모델 {args.model} / {args.size} / {args.quality}")
    made = skipped = failed = 0

    for eid, refs, scene in targets:
        out = EVENTS_DIR / f"{eid}.webp"
        if out.exists() and not args.force:
            print(f"건너뜀 {eid} (이미 존재)")
            skipped += 1
            continue
        print(f"생성 {eid} ... ", end="", flush=True)

        # 존재하는 레퍼런스만 첨부
        files, missing = [], []
        for r in refs:
            rp = ROOT / r
            (files if rp.exists() else missing).append(rp if rp.exists() else r)
        prompt = COMMON + scene
        try:
            t0 = time.time()
            if files:
                handles = [open(f, "rb") for f in files]
                try:
                    resp = client.images.edit(model=args.model, image=handles,
                                               prompt=prompt, size=args.size, quality=args.quality)
                finally:
                    for h in handles:
                        h.close()
            else:   # 레퍼런스가 하나도 없으면 일반 생성
                resp = client.images.generate(model=args.model, prompt=prompt,
                                               size=args.size, quality=args.quality)
            png = base64.b64decode(resp.data[0].b64_json)
            if args.keep_raw:
                (RAW_DIR).mkdir(parents=True, exist_ok=True)
                (RAW_DIR / f"{eid}.png").write_bytes(png)
            to_webp(png, out, args.width, args.wq)
            kb = out.stat().st_size // 1024
            warn = f"  ⚠ ref 누락: {', '.join(map(str, missing))}" if missing else ""
            print(f"✓ {out.relative_to(ROOT)} ({kb}KB, {time.time()-t0:.0f}s){warn}")
            made += 1
        except Exception as ex:
            print(f"✗ 실패: {ex}")
            failed += 1

    # EVENT_IMG 매핑 검증
    idx = (ROOT / "index.html").read_text(encoding="utf-8")
    registered = sum(1 for e in EVENTS if f"{e[0]}:'assets/events/{e[0]}.webp'" in idx.replace(" ", ""))
    print(f"EVENT_IMG 등록 확인: {registered}/{len(EVENTS)}개")
    print(f"완료 — 생성 {made}, 건너뜀 {skipped}, 실패 {failed}")


if __name__ == "__main__":
    main()
